sklearn/tree/_utils.pyx

```diff
@@ -25,17 +25,19 @@ np.import_array()
 # Helper functions
 # =============================================================================
 
-cdef realloc_ptr safe_realloc(realloc_ptr* p, size_t nelems) except *:
+cdef realloc_ptr safe_realloc(realloc_ptr* p, size_t nelems) nogil except *:
     # sizeof(realloc_ptr[0]) would be more like idiomatic C, but causes Cython
     # 0.20.1 to crash.
     cdef size_t nbytes = nelems * sizeof(p[0][0])
     if nbytes / sizeof(p[0][0]) != nelems:
         # Overflow in the multiplication
-        raise MemoryError("could not allocate (%d * %d) bytes"
-                          % (nelems, sizeof(p[0][0])))
+        with gil:
+            raise MemoryError("could not allocate (%d * %d) bytes"
```

Comment:
Okay. Thanks!

And I think as long as gil is acquired when we raise the error, it will propagate when `except *` construct is used... Infact raising cannot be done without gil for the same reason... *I think* this is why the whole `safe_realloc` was not made to release gil? I now changed it to acquire gil only when the memory is not sufficient and hence error has to be raised...